### PR TITLE
Configuration driven deployment scripts.

### PR DESCRIPTION
**Pull request by jplodge-pro**
## Reason
For `testnet` and `mainnet` environments, we will likely deploy to multiple networks. So, our deployer should be configuration driven to allow us to have 1 deployment mechanism that works with all deployment scenarios.

## Current Situation
The current deployment mechanism is somewhat involved. 
- We use `yarn` to call `make` to call `forge script`.  With each layer having requirements and assumptions.
- There are pre-scripts (`checkDb.js`) and post-scripts (`exporter.js`) embedded in the process.
- We context switch from `make`, to `node`, to `forge`, to `node` (and back again). (Actually unavoidable, I think)
- The 'dependency chain' between steps is non-existent (i.e. "you must invoke `yarn build` before `yarn deploy`" etc.)
- There are flat-file requirements, reflecting data in the database, used to perform contract existence checks and blockchain inspection.
- The Transaction Log inspection used to gather deployed contract details means contracts are stored against their Contract Name, so we can only deploy 1 of each contract. So, for instance, we could not deploy 2 `SimpleToken` instances reflecting different tokens.
- The deployment scripts are used within the test contracts, as a means of deploying test instances, with all the over-head of the flat files/chain inspection.

## Deployment
This is an attempt to codify our deployment process.
### Abstraction
We:
- define a **Deployment Unit** to be a collection of 1 or more contracts that are deployed to the target network.
- currently have 3 Deployment Units, called:
    - **Vaults**, deploying `CredbullFixedYieldVaultFactory`, `CredbullUpsideVaultFactory` and `CredbullWhiteListProvider`. 
    - **Vaults Support**, to support testing of **Vaults** contracts, deploying `SimpleToken` (for `$CBL`), `SimpleUSDC` (for `$USDC`) and `SimpleVault`. 
    - **CBL**, deploying the `CBL` token.
- have the following **Deployment Environments**:
    - **local**, for local development, deploying to an Anvil server running on __localhost__.
    - **test**, for unit testing purposes.
    - **ci**, for deploying to _Arbitrum Sepolia_ during the CI builds. 
    - **testnet**,  **TO BE CLARIFIED**.
    - **mainnet**, for deploying to Arbitrum **TO BE CLARIFIED**. 
- deploy to **Ethereum-based** networks and, as such, we will use the same addresses for all roles, across all networks.
    - Thus, our deployment configuration is mostly unchanged across networks.
- deploy using `forge script` with custom Solidity `Script`s.
- record all contracts deployed to any network within the **Supabase Database** associated with the environment. 
### Implementation
- There is 1 TOML configuration file per **Deployment Environment**, capturing the configuration for the **Deployment Units**.
    - The file is appropriately named: `environment.toml`
    - TOML is a human-readable format, so the contents of the configuration files are self-describing and not covered here.
- The deployments use `make` and are controlled from the `Makefile`.  They are also triggered via `yarn`.
- During deployment, there are checks to ensure the Deployment Unit contracts are not already deployed to the network. If deployed, the deployment fails.
- After deployment, the transaction log generated by `forge script` is introspected to capture the details of the deployed contracts, which is then uploaded to the **Supabase Database** instance for the **Deployment Environment**.

### Further Solutions
- Use `turbo` in all packages to enforce dependencies between build targets. This does not remove `yarn`. This could resolve the 'dependency chain' issue.
- `turbo` would also allow us to define input/output files sets, meaning only build where necessary, improving build times.
- No documentation on the steps needed to do each stage of development (added some) and deployment (outstanding).
- Our deployment scripts are a tangled weave of logic-branched deploy/test/chain query functionality. Embrace KISS and simplify them to do 1 thing (if possible)
- The manner in which we gate-keep contract deployments is pretty weak (the database table). But not sure there is a better way! :)  Are contract events subscription persistent? 
- Testing uses the `HelperConfig` and `DeployedContracts` to introspect the chain when there is simply no need.
    - In test, `setUp()` is run **per test case**, so each test runs in a clean environment (as is correct). Thus, we should create the contracts in `setUp()` and not look for previously deployed contracts. This applies even if the environment is not scraped clean per iteration. There is nothing  with deploying new contracts at different addresses per test case.

### CODE REVIEW

File: packages/contracts/script/deployer/Deployer.s.sol [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        vm.startBroadcast();
+        new CredbullFixedYieldVaultFactory(owner, operator, custodians);
+        new CredbullUpsideVaultFactory(owner, operator, custodians);
+        new CredbullWhiteListProvider(operator);
+        vm.stopBroadcast();
+    }
+}
+
+contract Arbitrum is Deployer {
```
Comment (jplodge-pro) [inline packages/contracts/script/deployer/Deployer.s.sol:43]: The idea here is that the command for deployment, to Arbitrum for example, can be:
```forge script srcipt/Deployer.s.sol:Arbitrum ...```

File: packages/contracts/resource/mainnet.toml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+[network.Arbitrum.deployment.token.cbl.actors.address]
+owner = "0xD79Be36f61fce3B8EF2FBF22b13B2b9a68eE15A2"
+minter = "0xaD3C004eE1f942BFDA2DA0D2DAaC94d6aC012F75"
+
+[network.Arbitrum.deployment.services.supabase]
+# NOTE (JL,2024-07-16): Given we upsert the deployed Contract Addresses to a Supabase instance, we need a Supabase
+#   instance PER DEPLOYMENT NETWORK!
+url = "https://kyvvhlnmoqibdihqrlmc.supabase.co"
+update_contract_addresses = true
```
Comment (lucasia) [inline packages/contracts/resource/mainnet.toml:29]: contract_addresses has a chain_id column.  so we can share the same supabase for different chains for that at least.  other tables may likewise need to be made chain aware.

File: packages/contracts/script/deployer/Configured.s.sol [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import { console2 } from "forge-std/console2.sol";
+import { StdChains } from "forge-std/StdChains.sol";
+import { stdToml } from "forge-std/StdToml.sol";
+
+/**
+ * @notice Any script that is based upon the multi-network syntax TOML Configuration should extend this.
+ *
+ * @dev This contract extends [CommonBase] to avoid directly extending [Script] or [Test], where it will be used.
+ */
+abstract contract Configured is CommonBase, StdChains {
```
Comment (jplodge-pro) [inline packages/contracts/script/deployer/Configured.s.sol:15]: This is the codified means of accessing the TOML configuration. There should be no other code that knows about TOML Config Keys.

File: packages/contracts/script/deployer/DeployVaults.s.sol [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * @notice The `forge script` invocation entrypoint, this deploys the Vaults Distribution Unit and, optionally,
+     *  the Vaults Support Distribution Unit, if deployment is enabled.
+     * @dev For command line invocation, the return values are ignored.
+     *
+     * @return fixedYieldVaultFactory The deployed [CredbullFixedYieldVaultFactory].
+     * @return upsideVaultFactory The deployed [CredbullUpsideVaultFactory].
+     * @return whiteListProvider The deployed [CredbullWhiteListProvider].
+     */
+    function run()
```
Comment (jplodge-pro) [inline packages/contracts/script/deployer/DeployVaults.s.sol:32]: I simplified these, removing all trace of 'test mode' or 'deploy checking'. Why? 
1. In a deployment setting, there is already a deployment check step, so having it IN the scripts is redundant.
2. In test, we don't actually care about existing contracts (`setUp()` re-set the chain each time) and must simply deploy.

File: packages/contracts/script/utils/exporter.js [hunk 5]
```diff
@@ -99,7 +98,9 @@ async function clearExistingData(client) {
 
 (async () => {
   try {
-    await exportAddress(loadConfiguration());
+    const chainId = Number(process.argv[2]);
```
Comment (jplodge-pro) [inline packages/contracts/script/utils/exporter.js:101]: Now accepts the Chain Id, passed in from the invoker (`make`, which can use a `cast chainid ${RPC_URL}`). This means the value is not in configuration and the scripts can be sensitive to the last invocation only.

File: packages/contracts/test/src/CredbullFixedYieldVaultTest.t.sol [hunk 6]
```diff
@@ -2,46 +2,51 @@
...
+
+contract CredbullFixedYieldVaultTest is VaultsSupportConfiguredTest {
+    uint256 private constant INITIAL_BALANCE = 1000;
+    uint256 private constant ADDITIONAL_PRECISION = 1e12;
+
+    address private alice = makeAddr("alice");
+    address private bob = makeAddr("bob");
 
-contract CredbullFixedYieldVaultTest is Test {
+    TestDeployVaults private vaultsDeployer;
```
Comment (jplodge-pro) [inline packages/contracts/test/src/CredbullFixedYieldVaultTest.t.sol:26]: Uses the Test Deployers, which load configuration from the `test.toml` file.

File: packages/contracts/test/src/CredbullFixedYieldVaultTest.t.sol [hunk 7]
```diff
@@ -2,46 +2,51 @@
...
 import { WhiteListProvider } from "@credbull/provider/whiteList/WhiteListProvider.sol";
 import { WhiteListPlugin } from "@credbull/plugin/WhiteListPlugin.sol";
 
-import { ParamsFactory } from "@test/test/vault/utils/ParamsFactory.t.sol";
+import { AltParamsFactory } from "@test/test/vault/utils/AltParamsFactory.t.sol";
 import { SimpleUSDC } from "@test/test/token/SimpleUSDC.t.sol";
+import { TestDeployVaults, TestDeployVaultsSupport } from "@test/test/deployer/TestDeployVaults.t.sol";
+import { VaultsSupportConfiguredTest } from "@test/test/deployer/ConfiguredTest.t.sol";
+
+contract CredbullFixedYieldVaultTest is VaultsSupportConfiguredTest {
```
Comment (jplodge-pro) [inline packages/contracts/test/src/CredbullFixedYieldVaultTest.t.sol:19]: Converted this test to use the changed deploy and configuration mechanisms.

File: packages/contracts/test/src/CredbullFixedYieldVaultTest.t.sol [hunk 8]
```diff
@@ -2,46 +2,51 @@
...
-        deployer = new DeployVaultFactory();
-        (,, whiteListProvider, helperConfig) = deployer.runTest();
-        params = new ParamsFactory(helperConfig.getNetworkConfig()).createFixedYieldVaultParams();
+        vaultsDeployer = new TestDeployVaults();
+        vaultsSupportDeployer = new TestDeployVaultsSupport();
+
+        (,, whiteListProvider) = vaultsDeployer.deploy();
+        (ERC20 cbl, ERC20 usdc,) = vaultsSupportDeployer.deploy();
+
+        params = new AltParamsFactory(usdc, cbl).createFixedYieldVaultParams();
```
Comment (jplodge-pro) [inline packages/contracts/test/src/CredbullFixedYieldVaultTest.t.sol:42]: The `ParamsFactory` really only needs access to the config and the deployed `CBL` and `USDC` tokens.

File: packages/contracts/test/src/vault/MaturityVaultTest.t.sol [hunk 9]
```diff
@@ -25,14 +25,16 @@ contract MaturityVaultTest is Test {
     uint256 private constant INITIAL_BALANCE = 1e6;
 
     function setUp() public {
-        helperConfig = new HelperConfig(true);
-        params = new ParamsFactory(helperConfig.getNetworkConfig()).createMaturityVaultParams();
+        deployer = new TestDeployVaultsSupport();
+        (ERC20 cbl, ERC20 usdc,) = deployer.deploy();
```
Comment (jplodge-pro) [inline packages/contracts/test/src/vault/MaturityVaultTest.t.sol:29]: Invokes the `deploy()` function to deploy the Vaults Support contracts, configured by the `test.toml`.

**Review by jplodge-pro — COMMENTED** [review]
Any questions?

**Comment by jplodge-pro** [comment]
Superseded by https://github.com/credbull/credbull-defi/pull/85

**Pull request closed without merging**